packages/core/contracts/financial-templates/perpetual-multiparty/PerpetualLiquidatable.sol

```diff
@@ -426,70 +435,77 @@ contract PerpetualLiquidatable is PerpetualPositionManager {
+            rewards.paidToLiquidator = _removeCollateral(rawLiquidationCollateral, rewards.payToLiquidator);
+            rewards.paidToSponsor = _removeCollateral(rawLiquidationCollateral, rewards.payToSponsor);
+            rewards.paidToDisputer = _removeCollateral(rawLiquidationCollateral, rewards.payToDisputer);
+
+            collateralCurrency.safeTransfer(liquidation.disputer, rewards.paidToDisputer.rawValue);
+            collateralCurrency.safeTransfer(liquidation.liquidator, rewards.paidToLiquidator.rawValue);
+            collateralCurrency.safeTransfer(liquidation.sponsor, rewards.paidToSponsor.rawValue);
+
             // In the case of a failed dispute only the liquidator can withdraw.
-        } else if (liquidation.state == Status.DisputeFailed && msg.sender == liquidation.liquidator) {
+        } else if (liquidation.state == Status.DisputeFailed) {
             // Pay LIQUIDATOR: collateral + dispute bond + returned final fee
-            withdrawalAmount = collateral.add(disputeBondAmount).add(finalFee);
-            delete liquidations[sponsor][liquidationId];
+            rewards.payToLiquidator = collateral.add(disputeBondAmount).add(finalFee);
+
+            // Transfer rewards and debit collateral
+            rewards.paidToLiquidator = _removeCollateral(rawLiquidationCollateral, rewards.payToLiquidator);
+
+            collateralCurrency.safeTransfer(liquidation.liquidator, rewards.paidToLiquidator.rawValue);
+
             // If the state is pre-dispute but time has passed liveness then there was no dispute. We represent this
             // state as a dispute failed and the liquidator can withdraw.
-        } else if (liquidation.state == Status.PreDispute && msg.sender == liquidation.liquidator) {
+        } else if (liquidation.state == Status.PreDispute) {
             // Pay LIQUIDATOR: collateral + returned final fee
-            withdrawalAmount = collateral.add(finalFee);
-            delete liquidations[sponsor][liquidationId];
+            rewards.payToLiquidator = collateral.add(finalFee);
+
+            // Transfer rewards and debit collateral
+            rewards.paidToLiquidator = _removeCollateral(rawLiquidationCollateral, rewards.payToLiquidator);
+
+            collateralCurrency.safeTransfer(liquidation.liquidator, rewards.paidToLiquidator.rawValue);
         }
-        require(withdrawalAmount.isGreaterThan(0), "Invalid withdrawal amount");
 
-        // Decrease the total collateral held in liquidatable by the amount withdrawn.
-        amountWithdrawn = _removeCollateral(rawLiquidationCollateral, withdrawalAmount);
+        // If the raw amounts of collateral to pay out are 0, which is possible if the liquidation is for 0 collateral,
+        // then revert because there are no rewards to pay.
+        require(
+            (rewards.payToLiquidator.add(rewards.payToSponsor).add(rewards.payToDisputer)).isGreaterThan(0),
+            "Rewards equal to 0"
+        );
```

Comment:
This may be necessary, but just playing devil's advocate, what downside does this guard against?

Reply:
It prevents the one edge case where liquidated collateral is 0 so rewards are 0 paid out. Regardless the liquidation will be deleted afterwards so maybe we can just remove this. The liquidator has no incentive to send out such a liquidation, in fact we can block liquidations for 0 collateral

Reply:
@mrice32 This [test](https://github.com/UMAprotocol/protocol/blob/master/packages/core/test/financial-templates/perpetual-multiparty/PerpetualLiquidatable.js#L1662) in particular show cases a situation where the `require` statement is hit.

I decided to block liquidations for 0 tokens in this [commit](https://github.com/UMAprotocol/protocol/pull/2152/commits/b6b774a3bf7ae516f72cfa61d8cdd9e49aebaf2e) because we explicitly block 0 collateral transactions for `deposit`, `withdraw`, and 0 token `create`'s, PTAL!

Reply:
Just to be totally clear here, once the liquidation is deleted, it will no longer get here because the withdrawable modifier will revert it, right? So this is _only_ guarding against 0-value liquidations, right?

Not super opinionated either way. Totally up to you. Might be worth removing if we get desperate for bytecode space :).

SGTM!

Reply:
> Just to be totally clear here, once the liquidation is deleted, it will no longer get here because the withdrawable modifier will revert it, right? So this is only guarding against 0-value liquidations, right?

Correct. This require statement was preventing the 1 `withdrawLiquidation()` call that could go through if someone liquidated 0 collateral.

Reply:
FWIW I did end up removing this require statement and also prevented the edge case by blocking 0 token l iqudiations